OpenRA.Game/Traits/World/MusicPlaylist.cs

```diff
@@ -0,0 +1,149 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System;
+using System.Linq;
+using OpenRA.GameRules;
+
+namespace OpenRA.Traits
+{
+	[Desc("Trait for music handling. Attach this to the world actor.")]
+	public class MusicPlaylistInfo : ITraitInfo
+	{
+		public readonly string StartingMusic = null;
+		public readonly bool LoopStartingMusic = false;
+
+		public object Create(ActorInitializer init) { return new MusicPlaylist(init.World, this); }
+	}
+
+	public class MusicPlaylist : INotifyActorDisposing
+	{
+		readonly MusicInfo[] random;
+		readonly MusicInfo[] playlist;
+
+		public readonly bool IsMusicAvailable;
+
+		MusicInfo currentSong;
+		bool repeat;
+
+		public MusicPlaylist(World world, MusicPlaylistInfo info)
+		{
+			IsMusicAvailable = world.Map.Rules.InstalledMusic.Any();
+
+			if (!IsMusicAvailable)
+			{
+				return;
+			}
+
+			playlist = world.Map.Rules.InstalledMusic.Select(a => a.Value).ToArray();
+
+			random = playlist.Shuffle(Game.CosmeticRandom).ToArray();
+
+			if (Game.Settings.Sound.MapMusic && !string.IsNullOrEmpty(info.StartingMusic) && world.Map.Rules.Music.ContainsKey(info.StartingMusic))
+			{
+				currentSong = world.Map.Rules.Music[info.StartingMusic];
+				repeat = info.LoopStartingMusic;
+			}
+			else
+			{
+				currentSong = Game.Settings.Sound.Shuffle ? random.First() : playlist.First();
+				repeat = Game.Settings.Sound.Repeat;
+			}
+
+			Play();
```

Comment:
This isn't correct, it completely ignores `Game.Settings.Sounds.MapMusic` (try starting the game with music enabled, then disable the checkbox, then switch mods). You'd need to do it like this:

```csharp
if (!Game.Settings.Sound.MapMusic)
    return;

if (!string.IsNullOrEmpty(info.StartingMusic) && ...)
    ...
else
    ...

Play()
```